src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -26,8 +26,10 @@ public static partial class JsonSerializer
         internal static bool ResolveMetadataForJsonObject(
             ref Utf8JsonReader reader,
             ref ReadStack state,
-            JsonSerializerOptions options)
+            JsonSerializerOptions options,
+            out string? referenceId)
         {
+            referenceId = null;
```

Comment:
Can we make this assignment only when returning `false` to have only one assignment in the happy path (not two)?

Reply:
After a second look, this could return `true` and be `null` on async scenarios if the stream runs out of data right before the end bracket `}`.

Then on the call-site, if the `TryCastPreservedValue` method throws, the `referenceId` parameter would be `null` and the exception message would be missing the id.

Maybe we must change the exception message to not include the reference id.